src/unit.rs

```diff
@@ -163,12 +160,69 @@ impl Unit {
     }
 }
 
-/// Perform a connection. Used recursively for redirects.
+/// Perform a connection. Follows redirects.
 pub(crate) fn connect(
-    unit: Unit,
+    mut unit: Unit,
+    use_pooled: bool,
+    mut body: SizedReader,
+) -> Result<Response, Error> {
+    let mut previous = vec![];
+    loop {
+        let mut resp = connect_inner(&unit, use_pooled, body, &previous)?;
+        previous.push(unit.url.to_string());
+        resp.history = previous;
+        // handle redirects
+        if (300..399).contains(&resp.status()) && unit.agent.config.redirects > 0 {
```

Comment:
Nit: I'd like to reorder the cases in this loop to favor early returns and reduce indentation. Specifically, something like:

```rust
if !(300..399).contains(&resp.status()) || unit.agent.config.redirects == 0 {
  return Ok(resp);
}
... otherwise, handle a redirect ...
```

This also allows getting rid of the `continue`, which IMO makes this feel more like a natural loop, where falling through to the bottom means the loop continues by default.

Reply:
Done, although I used `break` instead of return to avoid repeating the `Ok` a lot.

I also noticed that this returns `Ok(resp)` when `redirects == 0`, but `Err(TooManyRedirects)` for any other value of redirects. Is that intentional?

Reply:
Yep, that's intentional. `0` is a special value that means "turn off redirect handling."